lib/Language/Haskell/Stylish/Step/Imports.hs

```diff
@@ -160,6 +201,63 @@ formatImports maxCols options moduleStats rawGroup =
   forM_ group \imp -> printQualified options padNames stats imp >> newline
 
 
+--------------------------------------------------------------------------------
+groupAndFormat
+  :: Maybe Int
+  -> Options
+  -> ImportStats
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+  -> Editor.Edits
+groupAndFormat maxCols options moduleStats groups = edits
+  where edits = Editor.changeLines block (const regroupedLines)
+        regroupedLines =
+          intercalate [""] $
+          map (formatImports maxCols options moduleStats) grouped
+
+        grouped = groupByPatterns (groupPatterns options) imports
+        imports = concatMap toList groups
+
+        block = Block
+          (GHC.srcSpanStartLine . src $ head imports)
+          (GHC.srcSpanEndLine   . src $ last imports)
+        src = fromMaybe (error "regroupImports: missing location") .
+          GHC.srcSpanToRealSrcSpan . GHC.getLocA
+
+groupByPatterns
+  :: [Pattern]
+  -> [GHC.LImportDecl GHC.GhcPs]
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+groupByPatterns patterns allImports = go patterns allImports []
+  where
+    go [] [] groups            = groups
+    go [] imports groups       = groups <> [NonEmpty.fromList imports]
+    go (p : ps) imports groups =
+      let
+        (groups', rest) = extract p imports
+      in
+        go ps rest (groups <> groups')
+
+    extract p imports =
+      let
+        (matched, rest) =
+          partition (matches p) imports
+        subgroups =
+          groupBy ((==) `on` capture p) $ sortOn (capture p) matched
+      in
+        (map NonEmpty.fromList subgroups, rest)
```

Comment:
`NonEmpty.fromList` is partial! Are we sure there won't be any empty lists in there? `mapMaybe NonEmpty.nonEmpty` is I think the same but safe.

Reply:
In this case `subgroups` can't be empty by construction. Not sure it would be worth handling the empty case if it required several extra lines of code, but the `mapMaybe NonEmpty.nonEmpty` version is a great way to avoid the problem without that.

Reply:
Why can't `subgroups` be empty? What if `matched` is empty? Which it could be if you have a regex that doesn't match any of the imports?

Reply:
Sorry, I misspoke: `subgroups` can be empty, but it can't contain empty lists because `groupBy` will never generate empty lists. The `NonEmpty.fromList` call applies to the lists inside `subgroups`, not `subgroups` itself.

Reply:
In an ideal world we'd have `List.groupBy :: (a -> a -> Bool) -> [a] -> [NonEmpty a]`. But I also understand why we don't: converting from `NonEmpty a` to `[a]` is annoying, especially since it's logically a no-op. That's one of the big practical limitations I've found with H-M type systems: talking about *restrictions* on types adds a lot of distracting boilerplate which is not really necessary.

A refinement type system like LiquidHaskell avoids this problem and it would be pretty sweet to have that easily integrated into GHC just to reduce the cost of using things like `NonEmpty`. As-is, it's only worth using LiquidHaskell for projects that really care about proving additional properties about the code, but it requires too much setup just for convenience or a bit of additional type expressiveness.